add notes for n22778 (#513)
bitcoin-core-review-club/website: _posts/2022-06-08-#22778.md

---
layout: pr
date: 2022-06-08
title: "net processing: Reduce resource usage for inbound block-relay-only connections"
pr: 22778
authors: [jnewbery]
components: ["p2p"]
host: larryruane
status: upcoming

> commit: 9db82f1bca0bb51c2180ca4a4ad63ba490e79da4

Reviewer: ```suggestion
commit: 9db82f1
```

---

## Notes

> - (These notes describe behavior with or without this PR.)

Reviewer: I think you can remove this.

> - We are a `bitcoind` node. Each of our peers is either inbound (our

Reviewer: It's the Bitcoin Core PR Review Club, so there's no need to specify that our node is running bitcoind.
```suggestion
- Each of our node's peers is either inbound (our
```

  peer initiated the connection) or outbound (we initiated the connection).
  Once the connection is established, the relationship is symmetric (the
  protocol works the same in both directions).

- When we initiate an outbound connection, we may decide that we won't
  relay (forward) transactions to this peer.
  This peer is designated a `block-relay-only` peer. The peer

>   doesn't know that we're treating it as `block-relay-only`.

Reviewer: I don't think this description is accurate; we also don't relay addresses to block-relay-only connections.
Maybe pull in a description from and/or link to https://bitcoincore.reviews/19858 and https://bitcoincore.reviews/20726?

- When connecting with a peer, we allocate a `Peer` object.

>   If we'll need to notify and send transactions to
>   this peer (for example, if it's not a `block-relay-only` peer),

Reviewer: I think "need to" is too strong. And I think we usually say "announce" rather than "notify"
```suggestion
  If we plan to announce and send transactions to this peer,
```

  we create an additional `TxRelay` object to keep track
  of state needed for relaying transactions to this peer.

> - The `TxRelay` object contains (in simplified form):

Reviewer: Link to declaration?

   - the list of transactions (txids) that we know about
     but haven't yet notified this peer about
   - the "list" (actually a bloom filter) of transactions (txids) that
     the peer knows about (because either we sent the transaction
     to the peer, or it has sent it to us)
   - if we've enabled BIP37, the bloom filter that the peer sends us
     so that we can filter out transactions that it doesn't care about, 
  

> - If we advertise support for BIP37 (bloom filter) to our peer

Reviewer: ```suggestion
- If we advertise support for BIP37 bloom filters to our peer
```

  during version handshake, the peer can send us a bloom filter
  after which we won't send transactions that aren't present in
  the filter (to reduce bandwidth).

- If a node connects to us (it's one of our inbound connections), it
  can indicate that it doesn't want to receive transactions

>   from us. It does this using the `relay` field in the

Reviewer: I don't recommend linking to the wiki; it's often outdated and inaccurate. Linking to the code is better! https://github.com/bitcoin/bitcoin/blob/1b2e1d179c5b350cac69ee670e01355314f25e11/src/net_processing.cpp#L2763-L2764

  [version](https://en.bitcoin.it/wiki/Protocol_documentation#version)
  message.

- If we've advertised support for bloom filters to this peer,
  it may, at any time later, send us a `filterload` message,
  which has the side-effect of enabling transaction relay from
  us to this peer, if it had sent `relay = false` in the `version`
  message.

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or
   NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?

1. What's a bloom filter?

1. Why are some peers `block-relay-only`? Briefly, what's the purpose of having
   some peers being `block-relay-only`?

> 1. This PR reduces resource usage. Which resource, primarily? Why might this
>    be beneficial?

Reviewer: ```suggestion
1. This PR "reduces resource usage." Which resource is reduced, and by how much?

1. How does this PR reduce resource usage?
```

> 1. Why is the `TxRelay` object separate from the `Peer` object?

Reviewer: Please link to the declarations of these data structures. It helps specify exactly what you're referring to and gives attendees a hint for where to look in the codebase.

> 1. Why is `NODE_BLOOM` not the default behavior (why does a node choose

Reviewer: Link to https://github.com/bitcoin/bitcoin/pull/16152 for context?

   to advertise it, or not, using a service bit)?
   Why do most nodes apparently not enable this service?

> 1. Why would a peer send us `fRelay = false` in the version message?
> 
> 1. Would a peer send us `fRelay = false` if it _wasn't_ planning to send
>    us a `filterload` soon?

Reviewer: ```suggestion
1. Why would a peer send us `fRelay = false` in the version message?
Why might a peer send us `fRelay = false` if it _wasn't_ planning to send
   us a `filterload`?
```

---
layout: pr
date: 2022-06-08
title: "net processing: Reduce resource usage for inbound block-relay-only connections"
pr: 22778
authors: [jnewbery]
components: ["p2p"]
host: larryruane
status: upcoming
commit: 9db82f1
---

## Notes

- Each of our node's peers is either inbound (our
  peer initiated the connection) or outbound (we initiated the connection).
  Once the connection is established, the relationship is symmetric (the

>   protocol works the same in both directions).

Reviewer 2: To a certain extent, yes. The protocol is symmetric, but our treatment of nodes depends on whether the node is inbound or outbound. For example, we'll prefer fetch transactions or download blocks from outbound peers.

Reviewer: The latest change made it sound like preferential download from outbound is the *only* difference, so I've edited the wording a little bit in 001b263

- When we initiate an outbound connection, we may decide that we won't
  relay (forward) transactions or addresses to this peer.
  This peer is designated a `block-relay-only` peer.
  This kind of peer connection was covered in earlier review club meetings
  (PRs [15759](https://bitcoincore.reviews/15759),
  [19858](https://bitcoincore.reviews/19858),
  [20726](https://bitcoincore.reviews/20726)).
  The peer
  doesn't know that we're treating it as `block-relay-only`.

- When connecting with a peer, we allocate a
  [`Peer`](https://github.com/bitcoin/bitcoin/blob/1b2e1d179c5b350cac69ee670e01355314f25e11/src/net_processing.cpp#L203)

>   object.

Reviewer 2: Maybe add that we allocate a `CNode` object in net (connection layer), and a `Peer` object in net_processing (application layer).

  If we plan to announce and send transactions and addresses to
  this peer (for example, if it's not a `block-relay-only` peer),
  we create an additional

>   [`TxRelay`](https://github.com/bitcoin/bitcoin/blob/1b2e1d179c5b350cac69ee670e01355314f25e11/src/net_processing.cpp#L250)
>   object (linked from the `Peer` object) to track

Reviewer 2: Perhaps 
```suggestion
  [`TxRelay`](https://github.com/bitcoin/bitcoin/blob/1b2e1d179c5b350cac69ee670e01355314f25e11/src/net_processing.cpp#L250)
  struct (owned by the `Peer` object) to track
```

since TxRelay is a data structure rather than a class, and the relationship is that Peer owns the TxRelay struct.

  the state needed for relaying transactions to this peer.

- The `TxRelay` object contains (simplified):
   - the list of transactions (txids) that we know about
     but haven't yet announced to this peer

>    - the "list" (actually a bloom filter) of transactions (txids) that

Reviewer 2: ```suggestion
   - a rolling [bloom filter](https://en.wikipedia.org/wiki/Bloom_filter) of transactions (txids) that
```

I think referring to this as a list is confusing.

Author: I put the quotes around "list" to indicate that I'm speaking at a conceptual level, but I like your suggestion... Our audience will know that a bloom filter is conceptually a list (or actually a set).

     the peer knows about (because either we sent the transaction
     to the peer, or it has sent it to us)
   - if we've enabled BIP37, the bloom filter that the peer sends us
     (if and when it does)
     so that we can filter out transactions that it doesn't care about
  

> - If we advertise support for BIP37 bloom filters to our peer

Reviewer 2: Maybe link to https://github.com/bitcoin/bips/blob/b1791c24aa163eb6578d0bfaadcf44997484eeaf/bip-0037.mediawiki since the part about `fRelay` is important.

   among the commits?

**Bonus Questions**

1. This PR adds
   [`GetTxRelay()`](https://github.com/bitcoin/bitcoin/pull/22778/files#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R292)
   to safely (using a mutex) return a pointer to a peer's `TxRelay`
   structure. If the a call to `GetTxRelay()` (for example,
   [this one](https://github.com/bitcoin/bitcoin/pull/22778/files#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R1831-R1832))
   returns a non-null pointer, then the pointer is used without holding
   the mutex. What prevents the `TxRelay`
   structure from being deallocated just after the pointer to it is returned
   by `GetTxRelay()`, making the pointer invalid?

1. Some class variables are annotated `mutable`. What does this mean?
   _Hint:_ This is fairly common with mutex variables.

1. The pointer variable `tx_relay` is
   [initialized](https://github.com/bitcoin/bitcoin/pull/22778/files#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R2767)
   as `const`, but it's used to mutate the `TxRelay` structure it's pointing to.
   How is that possible?

## Meeting Log

{% irc %}

> {% endirc %}

Reviewer 2: Maybe comment this out until after the meeting.
